fiftyone/utils/beam.py

```diff
@@ -338,6 +343,130 @@ def beam_export(
             )
 
 
+def beam_map(
+    sample_collection,
+    map_fcn,
+    shard_size=None,
+    shard_method="id",
+    num_workers=None,
+    options=None,
+    progress=False,
+    verbose=False,
+):
+    """Applies the given function to each sample in the given collection via
+     `Apache Beam <https://beam.apache.org>`_ and saves any sample edits.
+
+    This function is a parallelized alternative to
+    :meth:`iter_samples(autosave=True) <fiftyone.core.collections.SampleCollection.iter_samples>`
+    that effectively performs the following operation in parallel::
+
+        for batch_view in fou.iter_batches(sample_collection, num_shards):
+            for sample in batch_view.iter_samples(autosave=True):
+                map_fcn(sample)
+
+    Example::
+
+        import fiftyone as fo
+        import fiftyone.utils.beam as foub
+        import fiftyone.zoo as foz
+
+        dataset = foz.load_zoo_dataset("cifar10", split="train")
+        view = dataset.select_fields("ground_truth")
+
+        def map_fcn(sample):
+            sample["gt_label"] = sample.ground_truth.label.upper()
+
+        foub.beam_map(view, map_fcn, num_workers=8, progress=True)
+
+    Args:
+        sample_collection: a
+            :class:`fiftyone.core.collections.SampleCollection`
+        map_fcn: a function to apply to each sample
+        shard_size (None): an optional number of samples to distribute to each
+            worker at a time. By default, samples are evenly distributed to
+            workers with one shard per worker
+        shard_method ("id"): whether to use IDs (``"id"``) or slices
+            (``"slice"``) to assign samples to workers
+        num_workers (None): the number of workers to use when no ``options``
+            are provided. The default is
+            :meth:`fiftyone.core.utils.recommend_process_pool_workers`
+        options (None): a
+            ``apache_beam.options.pipeline_options.PipelineOptions`` that
+            configures how to run the pipeline. By default, the pipeline will
+            be run via Beam's direct runner using multiprocessing with
+            ``num_workers`` workers
+        progress (False): whether to render progress bar(s) for each worker
+        verbose (False): whether to log the Beam pipeline's messages
+    """
+    if shard_method == "slice":
+        n = len(sample_collection)
+    else:
+        ids = sample_collection.values("id")
```

Comment:
I've noticed that this can be a bottleneck when the sample collection is large. Can we alternatively define bounds of _ids and avoid creating a huge list?

Reply:
Do you mean like how `shard_method="slice"` works?